pycloudlib/azure/instance.py

```diff
@@ -290,7 +304,82 @@ def add_network_interface(self) -> str:
 
     def remove_network_interface(self, ip_address: str):
         """Remove nic from running instance."""
-        raise NotImplementedError
+        # get ip address id
+        all_ips = list(self._network_client.public_ip_addresses.list_all())
+        ip_address_id = [
+            ip.id for ip in all_ips if ip.ip_address == ip_address
+        ]
+        if len(ip_address_id) > 0:
+            ip_address_id = ip_address_id[0]
+        else:
+            raise PycloudlibError(
+                f"IP address: {ip_address} not found in resources."
+            )
+        # Get details of the NICs attached to the VM.
+        vm_nics_ids = [
+            nic.id
+            for nic in self._instance["vm"].network_profile.network_interfaces
+        ]
+        all_nics: List[NetworkInterface] = list(
+            self._network_client.network_interfaces.list_all()
+        )
+        vm_nics = [nic for nic in all_nics if nic.id in vm_nics_ids]
+        primary_nic = [nic for nic in vm_nics if nic.primary][0]
+        nic_params = []
+        nic_to_remove: Optional[NetworkInterface] = None
+        for vm_nic in vm_nics:
+            nic_ip_address_id = vm_nic.ip_configurations[
+                0
+            ].public_ip_address.id  # type: ignore
+            if nic_ip_address_id == ip_address_id:
+                nic_to_remove = vm_nic
+            else:
+                nic_params.append({"id": vm_nic.id, "primary": vm_nic.primary})
+        if not nic_to_remove:
+            raise PycloudlibError(
+                f"Did not find NIC with ip address: {ip_address}"
+            )
+        # if primary nic is removed, then make the next NIC as primary
+        if nic_to_remove.primary:
+            nic_params[0]["primary"] = True
+            primary_nic = [
+                nic for nic in all_nics if nic.id == nic_params[0]["id"]
+            ][0]
+        self._remove_nic_from_vm(nic_params, primary_nic, all_ips)
```

Comment:
In all other implementations of this method, the NIC is also deleted, as it is right now, we do only effectively detach the interface and leak it.
Could we also implement a `self._remove_nic` or something in those lines?

Reply:
wrt deleting the nic, could we have a method for that, oorrr, we could pass an extra parameter to `remove_network_interface` to specify if we should delete it something like
```python
def remove_network_interface(self, ip_address, should_delete=False): ...
```
I am not sure about always deleting the NIC when detaching it from the VM.

Reply:
> I am not sure about always deleting the NIC when detaching it from the VM.
Why not?

That might be a good compromise, would you like that default to be `False` in azure? If yes, other clouds default to `True`, and would be a bit strange to have different defaults per clouds, or maybe not. Thoughts?

Reply:
Still thinking about this.

Reply:
Had a meeting with @aciba90 and decided to delete in this method and then later create methods for detaching without deleting.